### PR TITLE
Ndr/truck restrictions

### PR DESCRIPTION
**Pull request by nreinicke**
Adds in a new frontier model for truck specific restrictions like: maximum weight or maximum trailer length, adding a couple of new units along the way to support this.

Also ties the truck restrictions into the edge based rtree which should work for now but I think this raises the point again that the edge rtree should really have the frontier model in scope when it does an edge search instead of duplicating the process between the frontier model and the edge rtree search. To do this, it probably makes sense to elevate the rtrees into the search app (rather than being input plugins) and then we can easily provide the constructed frontier model to the rtree at query time.

### CODE REVIEW

**Comment by robfitzgerald** [comment]
> it probably makes sense to elevate the rtrees into the search app (rather than being input plugins) and then we can easily provide the constructed frontier model to the rtree at query time

totally. something like a MapMatchingService should be part of the CompassApp... aand, the FrontierModel would be passed into InputPlugin.process.
huh, that could be a little tricky, but, possible. CompassApp's SearchApp has the FrontierModelService but the FrontierModels still need to be built to be used. so, it's like we would want to call FrontierModelService.build() inside CompassApp.apply_input_plugins. 

building two frontier models per query is not ideal but should be cheap, it shouldn't take a ton of CPU/RAM. but it turns out to build only _one_, we would need to apply input plugins at the same time we are running search, and that breaks how we are chunking the jobs and could miss out on performance optimizations for chunking heterogeneous grid searches. `</rant>`

**Review by robfitzgerald — CHANGES_REQUESTED** [review]
looks good. i like the Weight + WeightUnits, thanks for putting that together and adding the sanity unit tests. i also think it's nice having the restrictions as an enum so the constraints and operations for each restriction type can take different shapes. i have two requests, both small, and then we could pull this in; but i have further thoughts for future work on this, included here:

### TruckParameter as an enum

i think we may want to come back around to this module and relax the design of the TruckParameters so that queries can provide a subset of the parameters, for example, when that information is not available for a vehicle. we could do this by turning `TruckParameter` into an enum and deserializing a `Vec<TruckParameter>` from the user query. if that were the case, we could even avoid the mandatory _O(n)_ sweep for each `Vec<TruckRestriction>` because we could match by type only the relevant restrictions. 

### this restriction model could absorb the other frontier models

if you think about it, the mechanism for these restrictions is the same as the turn restrictions and road class restrictions. we could just absorb them here:

```rust
pub enum TruckRestriction {
    MaximumTotalWeight((Weight, WeightUnit)),
    MaximumWeightPerAxle((Weight, WeightUnit)),
    MaximumLength((Distance, DistanceUnit)),
    MaximumWidth((Distance, DistanceUnit)),
    MaximumHeight((Distance, DistanceUnit)),
    MaximumTrailerLength((Distance, DistanceUnit)),
    RoadClass(u8),
    TurnRestriction(EdgeId, EdgeId)
}
```

then we generalize TruckParameter to RestrictionParameter. A RoadClass restriction parameter is a `RoadClass(HashSet(u8))` or whatever deserialized off the query. it would tidy things up since they all have the same logic

File: rust/routee-compass/src/app/compass/config/frontier_model/truck_restriction/mod.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+pub mod restriction;
+pub mod truck_parameters;
```
[the remark this answers is not in the document]
Comment (nreinicke) [reply rust/routee-compass/src/app/compass/config/frontier_model/truck_restriction/mod.rs:2]: Yeah that's a good idea

**Comment by nreinicke** [comment]
>  have further thoughts for future work on this, included here:

I like both of these suggestion and think they could make some good issues for future work

**Review by robfitzgerald — APPROVED** [review]
💁‍♂️

**Pull request merged**